src/main.rs

```diff
@@ -99,86 +102,160 @@ struct Args {
+        // let usage = BufferObjectFlags::SCANOUT | BufferObjectFlags::RENDERING;
+        let modifiers = std::iter::once(Modifier::Linear);
+        let gbm_surface: Surface<BufferObject<()>> = drm_display
+            .gbm_device
+            .create_surface_with_modifiers(drm_display.width, drm_display.height, format, modifiers)
+            .unwrap();
+
+        Window { gbm_surface, drm_display }
+    }
 
-    let buffer_data = vec![255u8; (width * height * 4) as usize];
+    fn swap_buffers(&self) {
+        // TODO(mbernat): move this elsewhere
+        let depth_bits = 24;
+        let bits_per_pixel = 32;
+
+        // SAFETY: this must be called exactly once after `eglSwapBuffers`, which happens in `frame.finish()`.
+        let buffer_object = unsafe { self.gbm_surface.lock_front_buffer().unwrap() };
+        // TODO(mbernat): We should recycle framebuffers;
+        // one can store an FB handle in buffer object's user_data() and reuse it when it exists
+        let fb = self
+            .drm_display
+            .gbm_device
+            .add_framebuffer(&buffer_object, depth_bits, bits_per_pixel)
+            .unwrap();
+        self.drm_display
+            .gbm_device
+            .set_crtc(
+                self.drm_display.crtc.handle(),
+                Some(fb),
+                (0, 0),
+                &[self.drm_display.connector.handle()],
+                Some(self.drm_display.mode),
+            )
+            .unwrap();
+        // TODO(mbernat): Subsequent mode setting should be done with gbm_device.page_flip()
+    }
+}
 
-    buffer_object.write(&buffer_data).unwrap().unwrap();
+mod rwh_impl {
+    use super::Window;
+    use gbm::AsRaw;
+    use raw_window_handle::*;
+
+    // SAFETY: surface is valid for the duration of the program
+    unsafe impl HasRawWindowHandle for Window {
+        fn raw_window_handle(&self) -> RawWindowHandle {
+            let mut handle = GbmWindowHandle::empty();
+            handle.gbm_surface = self.gbm_surface.as_raw() as *mut _;
+            RawWindowHandle::Gbm(handle)
+        }
+    }
 
-    let depth_bits = 32;
-    let bits_per_pixel = 32;
-    let fb = gbm.add_framebuffer(&buffer_object, depth_bits, bits_per_pixel).unwrap();
+    // SAFETY: device is valid for the duration of the program
+    unsafe impl HasRawDisplayHandle for Window {
+        fn raw_display_handle(&self) -> RawDisplayHandle {
+            let mut handle = GbmDisplayHandle::empty();
+            handle.gbm_device = self.drm_display.gbm_device.as_raw() as *mut _;
+            RawDisplayHandle::Gbm(handle)
+        }
+    }
+}
 
-    gbm.set_crtc(crtc_handle, Some(fb), (0, 0), &[connector.handle()], Some(preferred_mode))
-        .unwrap();
+fn main() {
+    let args = Args::parse();
+    let drm_display = DrmDisplay::init(&args).unwrap();
+    let window = Window::new(drm_display);
+    let glium_display = glutin::init(&window);
+    {
```

Comment:
The purpose was two-fold: not to pollute the scope with `glium` imports but also I initially thought frame dropping calls some additional logic. Turns out it doesn't and in any case this would be wrong because `window.swap_buffers()` would be called before the drop.

I'll remove it to prevent any further confusion.